File: internal/channels/classification.go
// classification.go — the RFC 0037 §A confidentiality classification lattice:
// the fixed, totally ordered four-level vocabulary every channel and (from the
// memory-substrate PRs on) every channel-derived memory entry is labeled with,
// plus the canonical rank helper the whole confidentiality boundary compares
// through. This file is the single Go-side source of the ordering; the Python
// twin is agents/persona_runtime/classification.py and the SQL-side form
// arrives with the §F recall filter (RFC 0037 PR 5). No code path compares
// level strings directly — every comparison goes through one of the helpers
// below.
//
// Fail-closed splits into THREE explicit rules (§A, revised 2026-07-19 —
// v0.3.12 review items 5/8), because "restrictive" flips direction across the
// helper's uses:
//
//	(a) stamping/labeling      — absent/unknown → `internal` (a channel the
//	    operator forgot to classify is confidential-by-default, never public)
//	(b) acting level at gate/  — absent/unknown → the `public` FLOOR (inject/
//	    recall time              return LESS; also closes the proto3 ""
//	                             version-skew window)
//	(c) entry protection level — unknown/unparseable → WITHHELD (treated as
//	    unknown/unparseable       above-`secret`: never injectable on a
//	                              corrupted label)
//
// A single blanket `unknown → internal` default would make (c)
// unimplementable through the helper — a corrupted entry label would rank
// `internal` and inject cleanly into any `internal` turn. So the core
// [classificationRank] takes only known levels (ok/sentinel return) and each
// rule is owned by exactly one named resolver — [RankForStamp] /
// [NormalizeForStamp] (a), [ActingRank] (b), [EntryRankOrWithhold] (c). No
// caller applies its own default.
package channels

import (
	"errors"
)

// Classification is an RFC 0037 §A confidentiality level. The vocabulary is
// fixed at the four constants below for v0.3.x (operator-defined lattice
// levels are RFC 0037 Open Question #1).
type Classification string

// The §A lattice, lowest to highest. The only operations the system needs
// are the total order (`a ≤ b`) and `max`, both taken over the ranks
// returned by [classificationRank].
const (
	// ClassificationPublic — rank 0. No confidentiality expectation; the
	// safe floor rule (b) resolves an unknown ACTING level to.
	ClassificationPublic Classification = "public"
	// ClassificationInternal — rank 1. Ordinary in-org conversation; the
	// default rule (a) stamps an unclassified channel/entry with.
	ClassificationInternal Classification = "internal"
	// ClassificationRestricted — rank 2. Sensitive; need-to-know within a
	// subset.
	ClassificationRestricted Classification = "restricted"
	// ClassificationSecret — rank 3. Highly sensitive; disclosure is a
	// material harm.
	ClassificationSecret Classification = "secret"
)

// DefaultClassification is the §A rule-(a) stamping default: what an
// absent-by-policy classification labels to. Deliberately `internal`, never
// `public` — a channel the operator forgot to classify is
// confidential-by-default.
const DefaultClassification = ClassificationInternal

// ErrInvalidClassification is returned by [Config.Validate] when an operator
// declares a classification outside the §A vocabulary. Load-time rejection is
// the belt-and-suspenders for the schema enum (`make validate`); the runtime
// resolvers never see the bad value.
var ErrInvalidClassification = errors.New("channels: invalid classification")

// The item-8 dark-window guard (DarkWindowMaxClassification /
// ErrClassificationAboveDarkWindow / CheckDarkWindowClassification) lived
// here from PR 1 and was DELETED in the PR that armed the §D hard gate
// (RFC 0037 PR 4), exactly as its REMOVAL note prescribed: with the gate,
// the gated recall_notes, and the §B single-channel-turn guard live, a
// declared level above `internal` is now an enforced boundary rather than
// an unkeepable promise. Operators may classify channels `restricted` /
// `secret` from v0.3.12 on.

// classificationRanks is the single Go-side source of the §A total order.
// Kept in lock-step with `CLASSIFICATION_RANKS` in
// agents/persona_runtime/classification.py — the cross-language agreement is
// pinned by identical literal tables in classification_test.go and
// tests/unit/python/test_classification.py, so a drift on either side fails
// that side's suite.
//
// The `schemas/channel.schema.json` enum is the third copy of this vocabulary.
// It is pinned to the PYTHON table (test_channel_config_schema.py derives its
// expected level list from `CLASSIFICATION_RANKS` rather than re-typing a
// literal), so this table's agreement with the schema is transitive: a level
// added here must be added to the Python table to pass the pin above, and that
// edit is what trips the schema test.
var classificationRanks = map[Classification]int{
	ClassificationPublic:     0,
	ClassificationInternal:   1,
	ClassificationRestricted: 2,
	ClassificationSecret:     3,
}

// classificationRank returns the §A lattice ordinal for a KNOWN level, and
// ok=false for anything else — including the empty string. Deliberately no
// default of any direction here: the three fail-closed rules disagree on what
// an unknown level means, so the default belongs to the named resolvers
// below, never to the core rank lookup (§A, the "restrictive flips direction"
// rationale in the file header).
//
// UNEXPORTED on purpose, so "each rule is owned by exactly one named resolver"
// is enforced by the compiler rather than by convention. An out-of-package
// gate author (PRs 4–5) can reach only [RankForStamp]/[NormalizeForStamp],
// [ActingRank], and [EntryRankOrWithhold] — each of which names the §A rule it
// applies — and cannot open-code a bare lookup plus a locally chosen default,
// which is the one mistake this file exists to prevent. Export it only
// alongside a fourth §A rule that genuinely needs it.
func classificationRank(level Classification) (rank int, ok bool) {
	rank, ok = classificationRanks[level]
	return rank, ok
}

// Valid reports whether level is one of the four §A levels. Comparison is
// exact — the vocabulary is lowercase and case-sensitive, matching the schema
// enum.
func (c Classification) Valid() bool {
	_, ok := classificationRanks[c]
	return ok
}

// NormalizeForStamp is rule (a) in the level domain: the classification to
// WRITE when stamping/labeling. A known level passes through; absent or
// unknown labels to [DefaultClassification] (`internal`), never `public`.
// This is what DM creation stamps from the `dm_default_classification` knob
// ([sqliteStore.GetOrCreateDM]) and what every future labeling path uses.
func NormalizeForStamp(level Classification) Classification {
	if level.Valid() {
		return level
	}
	return DefaultClassification
}

// RankForStamp is rule (a) in the rank domain: the ordinal of
// [NormalizeForStamp](level). Provided so stamp-side comparisons and the
// stamp-side write share one rule owner.
func RankForStamp(level Classification) int {
	rank, _ := classificationRank(NormalizeForStamp(level))
	return rank
}

// ActingRank is rule (b): the rank of the ACTING classification at gate/
// recall time. A known level ranks as itself; absent or unknown resolves to
// the `public` FLOOR — inject/return less. This is deliberately the opposite
// direction from [RankForStamp]: an event arriving with no classification
// (proto3 "" from an older orchestrator, an autonomous tick with no channel)
// must see the least-confidential view, not the `internal` default a
// stamp-side coercion would grant.
func ActingRank(level Classification) int {
	if rank, ok := classificationRank(level); ok {
		return rank
	}
	rank, _ := classificationRank(ClassificationPublic)
	return rank
}

// EntryRankOrWithhold is rule (c): the rank of a stored ENTRY protection
// level. A known level ranks as itself; unknown/unparseable returns ok=false
// — the entry is withheld, treated as above-`secret`, never coerced onto the
// lattice where it could inject on a corrupted label. Semantically this is
// the bare [classificationRank], named so gate-side callers state which §A
// rule they are applying instead of open-coding a lookup-plus-default that
// would silently pick a direction.
//
// Pure, and so is the Python twin `entry_rank_or_withhold` — rule (c)'s "and
// logged" half belongs to the CALLER in both languages, by design rather than
// by omission. The caller is the only layer holding the entry's identity (a
// bare `unknown protection_level "xyz"` cannot be triaged), and the §F recall
// filter calls this once per candidate row, so an in-helper warning would turn
// one corrupted batch into a log flood — loudest exactly when an operator is
// trying to read the gate's decisions. The security half is not delegated:
// withholding rides ok=false and cannot be forgotten without ignoring the
// result.
func EntryRankOrWithhold(level Classification) (rank int, ok bool) {
	return classificationRank(level)
}

File: agents/persona_runtime/classification.py
"""RFC 0037 §A classification lattice — the Python twin of the Go helper.

The fixed, totally ordered four-level confidentiality vocabulary every
channel (and, from the memory-substrate PRs on, every channel-derived
memory entry) is labeled with, plus the canonical rank helper the whole
confidentiality boundary compares through.  This module is the single
Python-side source of the ordering; the Go twin is
``internal/channels/classification.go`` and the SQL-side form arrives
with the §F recall filter (RFC 0037 PR 5).  No code path compares level
strings directly — every comparison goes through one of the helpers
below.

Fail-closed splits into THREE explicit rules (§A, revised 2026-07-19 —
v0.3.12 review items 5/8), because "restrictive" flips direction across
the helper's uses:

* **(a) stamping/labeling** — absent/unknown → ``internal`` (a channel
  the operator forgot to classify is confidential-by-default, never
  public).
* **(b) acting level at gate/recall time** — absent/unknown → the
  ``public`` FLOOR (inject/return LESS; also closes the proto3 ``""``
  version-skew window).
* **(c) entry protection level unknown/unparseable** — the entry is
  WITHHELD and logged (treated as above-``secret``: never injectable on
  a corrupted label).

A single blanket ``unknown → internal`` default would make (c)
unimplementable through the helper — a corrupted entry label would rank
``internal`` and inject cleanly into any ``internal`` turn.  So the core
:func:`_classification_rank` returns ``None`` for anything outside the
vocabulary, and each rule is owned by exactly one named resolver —
:func:`rank_for_stamp` / :func:`normalize_for_stamp` (a),
:func:`acting_rank` (b), :func:`entry_rank_or_withhold` (c).  No caller
applies its own default — and cannot: the core lookup is private on both
sides of the twin (``_classification_rank`` here, the unexported
``classificationRank`` in Go), so the only reachable doors are the three
that name the rule they apply.

**Where rule (c)'s "and logged" half lives: the CALLER, in both
languages.**  Every resolver here is pure — same contract as the Go twin's
``EntryRankOrWithhold``, deliberately, so the two do not disagree about
who owns the log line.  Two reasons the obligation sits with the caller
rather than in the helper:

* *Triage needs identity.*  The helper sees a bare label.  A gate-side
  ``WARNING`` naming the entry, its channel, and the acting level is
  actionable; ``unknown protection_level 'xyz'`` on its own is not.
* *Volume.*  The §F recall filter (RFC 0037 PR 5) evaluates this
  per candidate row, so an in-helper warning is one line per corrupted
  row — a single bad batch becomes a log flood, and the flood is loudest
  exactly when an operator is trying to read the gate's decisions.

The *security* half of rule (c) is not delegated: withholding is carried
by the ``None`` return and cannot be forgotten without ignoring the
result.  The log is observability layered on top, and every gate-side
caller (PRs 4–5) owes it.
"""

from __future__ import annotations

from typing import Final

# The §A lattice, lowest to highest.  The only operations the system needs
# are the total order and max, both taken over the ranks below.
CLASSIFICATION_PUBLIC: Final[str] = "public"
CLASSIFICATION_INTERNAL: Final[str] = "internal"
CLASSIFICATION_RESTRICTED: Final[str] = "restricted"
CLASSIFICATION_SECRET: Final[str] = "secret"

#: The §A rule-(a) stamping default: what an absent-by-policy classification
#: labels to.  Deliberately ``internal``, never ``public``.
DEFAULT_CLASSIFICATION: Final[str] = CLASSIFICATION_INTERNAL

#: The single Python-side source of the §A total order.  Kept in lock-step
#: with ``classificationRanks`` in ``internal/channels/classification.go`` —
#: the cross-language agreement is pinned by identical literal tables in
#: ``tests/unit/python/test_classification.py`` and the Go
#: ``classification_test.go``, so a drift on either side fails that side's
#: suite.
#:
#: The JSON-schema enum is the third copy of this vocabulary and is pinned to
#: THIS table, not to a literal of its own:
#: ``test_channel_config_schema.py`` derives its expected level list from here,
#: so adding or renaming a level without editing
#: ``schemas/channel.schema.json`` fails that suite.  Go ↔ schema agreement is
#: therefore transitive through this dict.
CLASSIFICATION_RANKS: Final[dict[str, int]] = {
    CLASSIFICATION_PUBLIC: 0,
    CLASSIFICATION_INTERNAL: 1,
    CLASSIFICATION_RESTRICTED: 2,
    CLASSIFICATION_SECRET: 3,
}


def _classification_rank(level: str | None) -> int | None:
    """Return the §A lattice ordinal for a KNOWN level, else ``None``.

    Deliberately no default of any direction here — including for
    ``None``/empty input: the three fail-closed rules disagree on what an
    unknown level means, so the default belongs to the named resolvers
    below, never to the core rank lookup (the "restrictive flips
    direction" rationale in the module docstring).  Comparison is exact —
    the vocabulary is lowercase and case-sensitive, matching the schema
    enum.

    PRIVATE on purpose, mirroring the Go twin's unexported
    ``classificationRank``: "each rule is owned by exactly one named
    resolver" should be a property of the module's surface, not a request
    in a docstring.  Importers get the three rule-named resolvers and
    :func:`is_valid_classification`; a gate author (PRs 4–5) has no
    unnamed lookup to open-code a locally chosen default against.  Only
    this module's own tests reach past the underscore.
    """
    if level is None:
        return None
    return CLASSIFICATION_RANKS.get(level)


def is_valid_classification(level: str | None) -> bool:
    """Report whether ``level`` is one of the four §A levels."""
    return level is not None and level in CLASSIFICATION_RANKS


def normalize_for_stamp(level: str | None) -> str:
    """Rule (a) in the level domain: the classification to WRITE when stamping.

    A known level passes through; absent or unknown labels to
    :data:`DEFAULT_CLASSIFICATION` (``internal``), never ``public``.
    """
    if level is not None and level in CLASSIFICATION_RANKS:
        return level
    return DEFAULT_CLASSIFICATION


def rank_for_stamp(level: str | None) -> int:
    """Rule (a) in the rank domain: the ordinal of ``normalize_for_stamp(level)``.

    Provided so stamp-side comparisons and the stamp-side write share one
    rule owner.  Total: ``normalize_for_stamp`` only returns known levels.
    """
    return CLASSIFICATION_RANKS[normalize_for_stamp(level)]


def acting_rank(level: str | None) -> int:
    """Rule (b): the rank of the ACTING classification at gate/recall time.

    A known level ranks as itself; absent or unknown resolves to the
    ``public`` FLOOR — inject/return less.  This is deliberately the
    opposite direction from :func:`rank_for_stamp`: an event arriving with
    no classification (proto3 ``""`` from an older orchestrator, an
    autonomous tick with no channel) must see the least-confidential view,
    not the ``internal`` default a stamp-side coercion would grant.
    """
    rank = _classification_rank(level)
    if rank is not None:
        return rank
    return CLASSIFICATION_RANKS[CLASSIFICATION_PUBLIC]


def acting_at_or_below_internal(level: str | None) -> bool:
    """The §C identity write-through bound: does the ACTING classification
    rank at or below ``internal``?

    ``True`` → the cross-room relationship write-through may proceed;
    ``False`` (a ``restricted``/``secret`` turn) → the caller falls back
    to a room-scoped note, which is stamped and gated.  Resolution is
    rule (b) — an absent/unknown acting level floors to ``public`` and
    proceeds, so every pre-classification path (tick, CLI, pre-v0.3.12
    producer) keeps its exact prior behaviour.  Named here — beside the
    rule owner it composes — so the one call site
    (:mod:`agents.tools.identity_write_through`) states the §A rule it
    applies instead of open-coding a rank comparison.
    """
    return acting_rank(level) <= CLASSIFICATION_RANKS[CLASSIFICATION_INTERNAL]


def entry_rank_or_withhold(level: str | None) -> int | None:
    """Rule (c): the rank of a stored ENTRY protection level.

    A known level ranks as itself; unknown/unparseable returns ``None`` —
    the entry is withheld, treated as above-``secret``, never coerced onto
    the lattice where it could inject on a corrupted label.

    Semantically this is the bare :func:`_classification_rank`, named so
    gate-side callers state which §A rule they are applying instead of
    open-coding a lookup-plus-default that would silently pick a
    direction.  Pure, exactly like the Go twin ``EntryRankOrWithhold``:
    rule (c)'s "and logged" half belongs to the caller, which is the only
    layer holding the entry's identity — see the module docstring for why,
    and note the recall filter calls this once per candidate row.
    """
    return _classification_rank(level)


def injectable_levels(acting: str | None) -> tuple[str, ...]:
    """Rules (b) + (c) in the LEVEL-SET domain: the entry protection levels
    injectable at the ACTING level ``acting`` — every vocabulary level whose
    rank is ``<= acting_rank(acting)``, rank-ascending.

    This is the shape the §D read-surface SQL predicates consume (RFC 0037
    PR 4 — the gated ``recall_notes``): the memory layer must not import
    this lattice (the import direction is ``persona_runtime → memory``), so
    the persona-side read surfaces resolve the acting level to this
    IN-list and pass it down as plain data.  A stored label OUTSIDE the
    returned set — including a corrupted one — falls out of the ``IN``
    predicate, which IS rule (c)'s withhold realised in SQL.  The SQL form
    is deliberately silent per row (the log-flood rationale in the module
    docstring); the Python-side gate over loaded rows owns the aggregated
    rule-(c) WARNING.

    Resolution of ``acting`` is rule (b): unknown/absent floors to
    ``public``, so an unbound turn's IN-list is ``("public",)`` — the
    least-disclosing set, never the stamp default.
    """
    bound = acting_rank(acting)
    return tuple(
        level for level, rank in CLASSIFICATION_RANKS.items() if rank <= bound
    )


def levels_below_stamp(level: str | None) -> tuple[str, ...]:
    """Rule (a) in the LEVEL-SET domain: the levels ranking STRICTLY below
    ``normalize_for_stamp(level)``, rank-ascending.

    The §C ``update_note`` re-stamp consumes this: an edit re-stamps a note
    to ``max(existing protection_level, acting L)`` — never lowers — and the
    memory layer realises that ``max`` as a raise-only SQL ``CASE`` gated on
    the existing value being IN this set (RFC 0037 PR 4).  A corrupted
    existing label is not in the set and therefore keeps its value,
    failing closed at read time per rule (c) rather than being silently
    laundered onto the lattice by an edit.

    Empty for ``level=None``/unknown-floor cases resolving to a ``public``
    stamp?  No — resolution here is rule (a) (absent → ``internal``), so the
    absent case returns ``("public",)``; only an explicit ``public`` acting
    level returns the empty tuple (nothing ranks below the floor).
    """
    bound = rank_for_stamp(level)
    return tuple(
        level_ for level_, rank in CLASSIFICATION_RANKS.items() if rank < bound
    )
